source_middleware.go

```diff
@@ -598,6 +576,179 @@ func (s *sourceWithSchemaContext) LifecycleOnDeleted(ctx context.Context, config
 	return s.Source.LifecycleOnDeleted(schema.WithSchemaContextName(ctx, s.contextName), config)
 }
 
+// -- SourceWithEncoding ------------------------------------------------------
+
+// SourceWithEncoding is a middleware that encodes the record payload and key
+// with the provided schema. The schema is registered with the schema service
+// and the schema subject is attached to the record metadata.
+type SourceWithEncoding struct{}
+
+func (s SourceWithEncoding) Apply(SourceMiddleware) {
+}
+
+func (s SourceWithEncoding) Wrap(impl Source) Source {
+	return &sourceWithEncoding{Source: impl}
+}
+
+// sourceWithEncoding is the actual middleware implementation.
+type sourceWithEncoding struct {
+	Source
+
+	keyWarnOnce     sync.Once
+	payloadWarnOnce sync.Once
+}
+
+func (s *sourceWithEncoding) Read(ctx context.Context) (opencdc.Record, error) {
+	rec, err := s.Source.Read(ctx)
+	if err != nil {
+		return rec, err
+	}
+
+	if err := s.encodeKey(ctx, &rec); err != nil {
+		return rec, err
+	}
+	if err := s.encodePayload(ctx, &rec); err != nil {
+		return rec, err
+	}
+
+	return rec, nil
+}
+
+func (s *sourceWithEncoding) encodeKey(ctx context.Context, rec *opencdc.Record) error {
+	if _, ok := rec.Key.(opencdc.StructuredData); !ok {
+		// log warning once, to avoid spamming the logs
+		s.keyWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record key is not structured, consider disabling the source schema key encoding using "%s: false"`, configSourceSchemaExtractionKeyEnabled)
```

Comment:
Well, there's no way to disable the encoding now, it always happens 😓 I think it would be good to still have the ability to skip this, if nothing else for performance reasons when you don't need schemas. That said, I wouldn't want us to add additional parameters for that, we already have parameters for enabling schema support in the other middleware. So then the question is, do we just use those? But that seems fishy, because this middleware would be using parameters that another middleware is declaring 😕

Reply:
> That said, I wouldn't want us to add additional parameters for that, we already have parameters for enabling schema support in the other middleware.

I guess you mean the middleware that extracts the schema?

But regarding disabling, yeah, this was a copy-paste from the extraction middleware. 

I agree that it would be good to make it possible to disable record data encoding. As for the params, extraction, and encoding are separate steps (and extraction might even be there), so separate params should be better IMO.

Reply:
Yes, I mean `sdk.schema.extract.payload.enabled` and `sdk.schema.extract.key.enabled`. My goal is to make this as simple as possible, as it's already complex, so I'm trying to see if we can avoid separate parameters.

Let's think this through. I'll only write about one parameter (`sdk.schema.extract.payload.enabled`), but the same logic applies to the key. The way it works in this PR - the value of `sdk.schema.extract.payload.enabled` is set to `true` by default, so the schema will be always extracted from the structured data. If a connector attaches its own schema (like `postgres` does), it would also have to disable this default setting, otherwise the schema would be extracted and overwritten. You're saying we should have a separate parameter for disabling only the encoding, but I'm thinking if the user ever wants to extract the schema _without_ encoding the data? I don't see when this would be the case, since we might lose precision when sending the structured data over gRPC to Conduit, it's not really a safe thing to do.

So what I'm proposing is to keep the behavior from the user's perspective and from the connector developer's perspective as it was. This means that `sdk.schema.extract.payload.enabled` is enabled by default, which means the schema gets extracted _unless_ the connector attached a schema already (extraction is a fallback). The encoding then only happens if a schema was attached, regardless if that was done by the middleware or the connector. The user can then disable the schema extraction and encoding using one switch.

Let me know if that makes sense.

Reply:
> You're saying we should have a separate parameter for disabling only the encoding, but I'm thinking if the user ever wants to extract the schema without encoding the data? 

Well, I think that if we want to make it possible to disable the encoding, then I think we should have a separate parameter.

Now, should we make it possible to disable the encoding? I agree, if a user wants the schema extracted, then he probably also wants the data encoded too (unless the situation is such that there's no risk of losing precision, and the schema is needed to create a destination collection). What if a connector is adding a schema on its own, and a user doesn't want it? I'm not sure if that's something that could actually happen. All in all, if we ever face such use cases, we can think about a solution, but for now, it looks like having a schema should imply a need to encode the data.

> The encoding then only happens if a schema was attached, regardless if that was done by the middleware or the connector.

Yup, it does make sense, and it's actually how it works now (the above lines were only a copy-paste for logging purposes), so I can just leave it, and clarify the log message.